src/parsers/simple.rs

```diff
@@ -1,294 +1,442 @@
-    s: sign ~
-    h: hour ~
-    m: empty_or!(
-        chain!(
-            tag!(":")? ~ m: minute , || { m }
-        ))
-    ,
-    || { (Some(s * (h as i32)), Some(s * (m.unwrap_or(0) as i32))) }
-));
-
-named!(timezone_utc <(Option<i32>,Option<i32>)>, map!(tag!("Z"), |_| (Some(0),Some(0))));
-
-// named!(pub parse_datetime <GedcomxDate>, chain!(d:datetime, || GedcomxDate::Simple(d)));
-
-named!(pub simple_date <GedcomxDate>, chain!(
-    a: approximate? ~
-    d: datetime, || {
+            tz_offset_minutes: tz.1,
+        },
+    ))
+}
+
+fn sign(i: &str) -> IResult<&str, i32> {
+    alt((tag("-"), tag("+")))(i).map(|(remaining, s)| {
+        (
+            remaining,
+            match s {
+                "+" => 1,
+                _ => -1,
+            },
+        )
+    })
+}
+
+fn timezone_minute(i: &str) -> IResult<&str, i32> {
+    let (remaining, (_, m)) = (opt(tag(":")), opt(minute)).parse(i)?;
+    Ok((remaining, m.unwrap_or(0) as i32))
+}
+
+fn timezone_hour(i: &str) -> IResult<&str, (Option<i32>, Option<i32>)> {
+    let (remaining, (s, h, m)) = (sign, hour, timezone_minute).parse(i)?;
+    Ok((remaining, (Some(s * (h as i32)), Some(s * m))))
+}
+
+fn timezone_utc(i: &str) -> IResult<&str, (Option<i32>, Option<i32>)> {
+    map(tag("Z"), |_| (Some(0), Some(0)))(i)
+}
+
+pub fn simple_date(i: &str) -> IResult<&str, GedcomxDate> {
+    let (remaining, (a, d)) = (opt(approximate), datetime).parse(i)?;
+    Ok((
+        remaining,
         GedcomxDate::Simple(Simple {
             date: d.date,
             time: d.time,
-            approximate: match a { Some(_) => true, None => false}
-        })
-    }
-));
-
-named!(pub datetime <DateTime>, chain!(
-    d: parse_date ~
-    t: opt!(complete!(chain!(tag!("T") ~ time: parse_time, || time)))
-    ,
-    || {
-        DateTime{
-            date: d,
-            time: t,
-        }
-    }
-));
+            approximate: a.is_some(),
+        }),
+    ))
+}
+
+fn time_piece(i: &str) -> IResult<&str, Time> {
+    let (remaining, (_, t)) = (tag("T"), parse_time).parse(i)?;
+    Ok((remaining, t))
+}
+
+pub fn datetime(i: &str) -> IResult<&str, DateTime> {
+    let (remaining, (d, t)) = (parse_date, opt(time_piece)).parse(i)?;
+    Ok((remaining, DateTime { date: d, time: t }))
+}
 
 #[cfg(test)]
 mod tests {
-
-    use nom::IResult::*;
+    use nom::error::Error;
 
     use super::super::super::{Date, Time};
-    use super::{ymd, parse_time};
-    use super::{year, month, day};
+    use super::{day, month, year};
     use super::{hour, minute, second};
+    use super::{parse_time, ymd};
 
     #[test]
     fn test_year() {
-        assert_eq!(Done(&[][..], 2015), year(b"+2015"));
-        assert_eq!(Done(&[][..], -333), year(b"-0333"));
-        assert_eq!(Done(&b"-"[..], 2015), year(b"+2015-"));
-
-        assert!(year(b"2003").is_err());
-        assert!(year(b"+abcd").is_err());
-        assert!(year(b"+2a03").is_err());
-
-        assert!(year(b"+203").is_incomplete());
```

Comment:
This returned incomplete because in nom 1 there wasn't consistent separation between streaming and complete versions of the combinators, so this turned out to be a streaming one. We instead get error now since incomplete should never be returned from a complete combinator.